odoo/addons/base/models/ir_property.py

```diff
@@ -30,92 +30,81 @@
-    res_id = fields.Char(string='Resource', index=True, help="If not set, acts as a default value for new resources",)
+    res_id = fields.Integer(string='Resource', index=True, help="If 0, acts as a default value for new resources", required=True, default=0)
     company_id = fields.Many2one('res.company', string='Company', index=True)
+
     fields_id = fields.Many2one('ir.model.fields', string='Field', ondelete='cascade', required=True, index=True)
+    name = fields.Char(store=True, related='fields_id.name', index=True)
+    type = fields.Selection(store=True, readonly=True, related='fields_id.ttype', index=True)
+    model = fields.Char(store=True, readonly=True, related='fields_id.model', index=True)
+
     value_float = fields.Float()
     value_integer = fields.Integer()
     value_text = fields.Text()  # will contain (char, text)
     value_binary = fields.Binary()
-    value_reference = fields.Char()
     value_datetime = fields.Datetime()
-    type = fields.Selection([('char', 'Char'),
-                             ('float', 'Float'),
-                             ('boolean', 'Boolean'),
-                             ('integer', 'Integer'),
-                             ('text', 'Text'),
-                             ('binary', 'Binary'),
-                             ('many2one', 'Many2One'),
-                             ('date', 'Date'),
-                             ('datetime', 'DateTime'),
-                             ('selection', 'Selection'),
-                             ],
-                            required=True,
-                            default='many2one',
-                            index=True)
 
-    @api.multi
-    def _update_values(self, values):
-        value = values.pop('value', None)
-        if not value:
-            return values
-
-        prop = None
-        type_ = values.get('type')
-        if not type_:
-            if self:
-                prop = self[0]
-                type_ = prop.type
-            else:
-                type_ = self._fields['type'].default(self)
-
-        field = TYPE2FIELD.get(type_)
-        if not field:
-            raise UserError(_('Invalid type'))
-
-        if field == 'value_reference':
-            if isinstance(value, models.BaseModel):
-                value = '%s,%d' % (value._name, value.id)
-            elif isinstance(value, int):
-                field_id = values.get('fields_id')
-                if not field_id:
-                    if not prop:
-                        raise ValueError()
-                    field_id = prop.fields_id
+    value = MysteryField(compute='_compute_value', inverse='_set_value')
+
+    @api.depends('type', 'value_float', 'value_integer', 'value_text', 'value_binary', 'value_datetime', 'fields_id.relation')
+    def _compute_value(self):
+        for prop in self:
+            prop.value = False
+            prop_type = prop.type
+
+            if prop_type in ('char', 'text', 'selection'):
```

Comment:
@xmo-odoo Have you considered a simple dispatcher on this? Or even just ordering the if statements by the most common request in largish databases - which if I had to guess would be float then m2o (product pricing, then account/journal ids. I don't know, maybe property fields are no longer a performance issue, but they used to be, especially once you get a lot of products because of all those floats in there. Even a smallish setup and its close to a million of them. thats only 50,000 products over 10 companies.